File: scripts/algorithms/Binary Tree Zigzag Level Order Traversal/Binary Tree Zigzag Level Order Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        Queue<TreeNode> q=new LinkedList<TreeNode>();
        List<List<Integer>> res=new ArrayList<>();
        if(root==null){
            return res;
        }
        q.offer(root);
        boolean flag=true;
        while(!q.isEmpty()){
            int size=q.size();
            List<Integer> curr=new ArrayList<>();
            for(int i=0;i<size;i++){
                if(q.peek().left!=null){
                    q.offer(q.peek().left);
                }
                if(q.peek().right!=null){
                    q.offer(q.peek().right);
                }
                if(flag==true){
                    curr.add(q.poll().val);
                }else{
                    curr.add(0,q.poll().val);
                }
            }
            if(flag==true){
                flag=false;
            }else{
                flag=true;
            }
            res.add(curr);
        }
        return res;
    }
}

File: scripts/algorithms/B/Binary Tree Zigzag Level Order Traversal/Binary Tree Zigzag Level Order Traversal.cpp
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
    vector<vector<int>> ans;
    if(!root) return ans;
    vector<int> r;
    queue<TreeNode*> q;
    int count=1;
    q.push(root);
    q.push(NULL);
    while(1){
        if(q.front()==NULL){
            q.pop();
            if(count%2==0) reverse(r.begin(),r.end());
            ans.push_back(r);
            if (q.empty()) return ans;
            q.push(NULL);
            r.resize(0);
            count++;
            continue;
        }
        
            r.push_back((q.front())->val);
            if(q.front()->left)q.push((q.front())->left); 
            if(q.front()->right)q.push((q.front())->right);
            
            q.pop();
    }
    return ans;

    }
};
